sns/parity_math.c

```diff
@@ -500,29 +551,110 @@ static void reed_solomon_encode(struct m0_parity_math *math,
 
 #undef PARITY_MATH_REGION_ENABLE
 }
+#endif /* __KERNEL__ */
+
+#ifndef __KERNEL__
+static void isal_encode(struct m0_parity_math *math,
+			const struct m0_buf *data,
+			struct m0_buf *parity)
+{
+	uint8_t **data_frags = NULL;
+	uint8_t **parity_frags = NULL;
+	uint32_t  i;
+	uint32_t  data_count;
+	uint32_t  parity_count;
+	uint32_t  block_size;
+
+	M0_PRE(math != NULL);
+	M0_PRE(data != NULL);
+	M0_PRE(parity != NULL);
+
+	M0_ENTRY();
+
+	data_count = math->pmi_data_count;
+	parity_count = math->pmi_parity_count;
+	block_size = data[0].b_nob;
+
+	M0_ALLOC_ARR(data_frags, data_count);
+	if (data_frags == NULL) {
+		M0_ERR_INFO(-ENOMEM, "failed to allocate memory for array of "
+			    "data fragments");
+		goto fini;
+	}
+
+	M0_ALLOC_ARR(parity_frags, parity_count);
+	if (parity_frags == NULL) {
+		M0_ERR_INFO(-ENOMEM, "failed to allocate memory for array of "
+			    "parity fragments");
+		goto fini;
+	}
+
+	data_frags[0] = (uint8_t *)data[0].b_addr;
+	for (i = 1; i < data_count; ++i) {
+		if (block_size != data[i].b_nob) {
+			M0_ERR_INFO(-errno, "data block size mismatch. "
+				    "block_size = %u, data[%u].b_nob=%u",
+				    block_size, i, data[i].b_nob);
+			goto fini;
+		}
+		data_frags[i] = (uint8_t *)data[i].b_addr;
+	}
+
+	for (i = 0; i < parity_count; ++i) {
+		if (block_size != parity[i].b_nob) {
+			M0_ERR_INFO(-errno, "parity block size mismatch. "
+				    "block_size = %u, parity[%u].b_nob=%u",
+				    block_size, i, parity[i].b_nob);
+			goto fini;
+		}
+		parity_frags[i] = (uint8_t *)parity[i].b_addr;
+	}
+
+	M0_LOG(M0_DEBUG, "initialize tables for fast Erasure Code encode.");
+	ec_init_tables(data_count, parity_count,
+		       &math->encode_matrix[data_count * data_count],
+		       math->g_tbls);
```

Comment:
Can this initialization be moved into "m0_parity_math_init()"?
Or this is needed for every ec_encode_data() call?

Reply:
We need g_tbls for encoding / decoding i.e. in ec_encode_data(). While decoding, the parameters passed to initialize the tables are different from those used in encoding. So to ensure that we have correct values in tables, it required to initialize it before calling ec_encode_data().

Reply:
@sanjognaik Can we use 2 pointers in that case one to be used for encode and other for decode with the encode/decode routines calling the appropriate pointers. This is because we should avoid to initialize things during runtime if possible.

Reply:
Created separate pointer 'pmi_encode_tbls' for encoding and initialized it in m0_parity_math_init() function now. Will use separate pointer for decode as well in PR for recover function.